File: dashboard/internal/paths/paths.go
// Package paths is the dashboard's single resolver for user-data locations,
// mirroring lib/paths.mjs. User data (config/, data/, output/, transcripts/)
// lives in <repo>/user — its own gitignored git repo — unless CAREER_OPS_USER_DIR points
// elsewhere. Paths stored inside user data (tracker report links such as
// `data/reports/…`) are relative to the user dir: resolve them with User.
package paths

import (
	"os"
	"path/filepath"
	"strings"
)

// UserDir returns the user-data root for the career-ops repo at repo.
func UserDir(repo string) string {
	if d := os.Getenv("CAREER_OPS_USER_DIR"); d != "" {
		if abs, err := filepath.Abs(d); err == nil {
			return abs
		}
		return d
	}
	return filepath.Join(repo, "user")
}

// User joins rel (e.g. a stored tracker link) onto the user-data root.
func User(repo string, rel ...string) string {
	return filepath.Join(append([]string{UserDir(repo)}, rel...)...)
}

// Config joins rel onto the user config/ directory.
func Config(repo string, rel ...string) string {
	return User(repo, append([]string{"config"}, rel...)...)
}

// Data joins rel onto the user data/ directory.
func Data(repo string, rel ...string) string {
	return User(repo, append([]string{"data"}, rel...)...)
}

// Output joins rel onto the user output/ directory.
func Output(repo string, rel ...string) string {
	return User(repo, append([]string{"output"}, rel...)...)
}

// Display renders abs for agent prompts that run with cwd=repo: repo-relative
// when abs is inside the repo, otherwise absolute.
func Display(repo, abs string) string {
	rel, err := filepath.Rel(repo, abs)
	if err != nil || rel == "." || strings.HasPrefix(rel, "..") {
		return abs
	}
	return rel
}

File: lib/paths.mjs
// paths.mjs — the single resolver for user-data locations.
//
// User data (config/, data/, output/, transcripts/) lives in its own git repo at
// `user/`, gitignored by this repo. Set CAREER_OPS_USER_DIR to point at a
// different checkout. Every script resolves user paths through this module —
// never join the repo root with 'config'/'data'/'output'/'transcripts'.
//
// Paths stored INSIDE user data (tracker report links `[NUM](data/reports/…)`,
// TSV link columns, trace/review JSON) are relative to USER_DIR, so the data
// repo is self-contained; resolve them with userPath(). Code comments that say
// `data/jds/…` etc. mean USER_DIR-relative.
import { dirname, isAbsolute, join, relative, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

export const REPO_DIR = dirname(dirname(fileURLToPath(import.meta.url)));
export const USER_DIR = process.env.CAREER_OPS_USER_DIR
  ? resolve(process.env.CAREER_OPS_USER_DIR)
  : join(REPO_DIR, 'user');

export const CONFIG_DIR = join(USER_DIR, 'config');
export const DATA_DIR = join(USER_DIR, 'data');
export const OUTPUT_DIR = join(USER_DIR, 'output');
export const TRANSCRIPTS_DIR = join(USER_DIR, 'transcripts');

export const APPLICATIONS_FILE = join(DATA_DIR, 'applications.md');
export const JDS_DIR = join(DATA_DIR, 'jds');
export const REPORTS_DIR = join(DATA_DIR, 'reports');
export const TRACKER_ADDITIONS_DIR = join(DATA_DIR, 'tracker-additions');
export const SCAN_HISTORY_DB = join(DATA_DIR, 'scan-history.db');

/** Absolute path for a USER_DIR-relative path (e.g. a stored tracker link). */
export const userPath = (...rel) => join(USER_DIR, ...rel);

/** Path for messages and agent prompts: repo-relative inside the repo, else absolute. */
export const displayPath = (abs) => {
  const rel = relative(REPO_DIR, abs);
  return rel && !rel.startsWith('..') && !isAbsolute(rel) ? rel : abs;
};
